[12.0][ADD] stock_putaway_storage_type_strategy

grindtildeath opened the pull request.

grindtildeath committed:
[ADD] stock_putaway_storage_type_strategy

grindtildeath committed:
[DROPME] oca_dependencies.txt

guewen commented on stock_putaway_storage_type_strategy/demo/location_storage_type.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
```
```suggestion
<odoo noupdate="1">
```

guewen commented on stock_putaway_storage_type_strategy/demo/stock_putaway_rule.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<odoo>
```
```suggestion
<odoo noupdate="1">
```

guewen commented on stock_putaway_storage_type_strategy/models/location_storage_type.py:
```diff
@@ -0,0 +1,74 @@
+    do_not_mix_lots = fields.Boolean(
+        help="If checked, moves to the destination location will only be "
+             "allowed if the location contains product of the same "
+             "stock.production.lot"
```
```suggestion
             "lot."
```

guewen commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,160 @@
+
+    @api.depends(
+        'child_ids', 'location_id', 'location_id.pack_storage_strategy'
+    )
```
You should check `location_id.display_pack_storage_strategy` as with more than one level, the current depends would not detect the change.

Actually, I'm not sure we should hide the field on locations without children: when creating locations, you'll start by creating the zone/area, configure the strategy, then create the children locations. At worse, setting the strategy on a leaf will have no effect.

grindtildeath replied:
> setting the strategy on a leaf will have no effect

This is actually why I wanted to hide the field as I don't think it make sense to display something meaningless... :man_shrugging:

guewen replied:
It has meaning when you are creating the parent before the children (at this point, the parent is a leaf).

guewen commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,160 @@
+            storage_locations = dest_location.get_storage_locations(
+                product=product
+            )
+            allowed_location = storage_locations.select_allowed_location(
```
I would rename it `select_first_allowed_location` for clarity.

guewen commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,160 @@
+                    allowed_location_storage_types |= location_storage_type
+            elif location_storage_type.do_not_mix_products:
+                if location_storage_type.do_not_mix_lots:
+                    quant = self.env.context.get('putaway_quant')
```
Shouldn't we pass the quant around as a parameter in the `select_allowed_location` and `filter_restrictions` methods?

rousseldenis replied:
Put this at the beginning

guewen reviewed (approved):
A couple of details but globally LGTM

rousseldenis commented on stock_putaway_storage_type_strategy/models/location_storage_type.py:
```diff
@@ -0,0 +1,74 @@
+        help="If checked, moves to the destination location will only be "
+             "allowed if the location contains the same product."
+    )
+    # max_height = fields.Boolean(
```
Remove if not necessary

rousseldenis commented on stock_putaway_storage_type_strategy/models/location_storage_type.py:
```diff
@@ -0,0 +1,74 @@
+    @api.constrains('only_empty', 'do_not_mix_lots', 'do_not_mix_products')
+    def _check_empty_mix(self):
+        for location_storage_type in self:
+            if (
```
I like
```suggestion
            if any(location_storage_type.only_empty and (location_storage_type.do_not_mix_lots or 
                location_storage_type.do_not_mix_product) for location_storage_type in self):
```

 to avoid an indentation, but that's me.

rousseldenis commented on stock_putaway_storage_type_strategy/models/location_storage_type.py:
```diff
@@ -0,0 +1,74 @@
+        if not self.do_not_mix_products:
+            self.do_not_mix_lots = False
+
+    # @api.constrains('max_height', 'max_height_value', 'max_height_uom')
```
Remove this if not necessary

rousseldenis commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,160 @@
+        for location in self:
+            # Do not display on locations without children
+            if not location.child_ids:
+                location.display_pack_storage_strategy = False
```
No need of affectation (for < v13), that's default

rousseldenis commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,160 @@
+            if lot is not None:
+                extra_domain = OR([extra_domain, [('lot_id', '!=', lot.id)]])
+            domain = AND([domain, extra_domain])
+        return self.env['stock.quant'].search(domain, limit=1)
```
I would have built the domain in an other function, to make search(_prepare_domain_function()) and ease override

rousseldenis commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,160 @@
+            if lot is not None:
+                extra_domain = OR([extra_domain, [('lot_id', '!=', lot.id)]])
+            domain = AND([domain, extra_domain])
+        return self.env['stock.move.line'].search(domain, limit=1)
```
Same as above

rousseldenis commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,160 @@
+    def filter_restrictions(self, matching_location_storage_types, product=None):
+        allowed_location_storage_types = self.env['stock.location.storage.type']
+        for location_storage_type in matching_location_storage_types:
+            if location_storage_type.only_empty:
```
Should these 'strategies' be called using built functions with getattr('xxx_only_empty') to ease override ?

rousseldenis reviewed (commented):
Some remarks.

grindtildeath committed:
Apply suggestions from code review

Co-Authored-By: Guewen Baconnier <guewen.baconnier@camptocamp.com>

grindtildeath committed:
fixup! [ADD] stock_putaway_storage_type_strategy

grindtildeath committed:
fixup! [DROPME] oca_dependencies.txt

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+    _inherit = 'stock.location'
+
+    pack_storage_strategy = fields.Selection(
+        selection=[('none', 'None'),
```
We need to be consistent with the concept of storage kind here: https://github.com/OCA/stock-logistics-warehouse/pull/711/files#diff-bf18e95ba21ff1ac092947117a88571aR4

This should only be defined on location of kind "area".

There must also be a chaotic strategy that randomly select a location.

guewen replied:
It doesn't seem mandatory to me to bind this put-away concept to the location kind concept.
Could be a glue module to make the "UI-sugar" between them.

jbaudoux replied:
Regarding my comment on chaotic, in fact we should have in this module the chaotic strategy and in https://github.com/OCA/wms/pull/6 only the ABC preference. So we should be able to implement chaotic without ABC.

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+        self.ensure_one()
+        locations = self.browse()
+        if self.pack_storage_strategy == 'none':
+            locations = self
```
We need here a random picker for chaotic storage on children locations

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+                    package_storage_type, quant, product=product
+                )
+            ):
+                return dest_location
```
If strategy is none and location does not allow package, then jump to next rule otherwise you retest the same.
Nevertheless, we do not need this anymore following my comment just above in the code.

grindtildeath committed:
Add stock_quant_package_product_packaging as dependency

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+
+    pack_storage_strategy = fields.Selection(
+        selection=[('none', 'None'),
+                   ('ordered_locations', 'Ordered Children Locations')],
```
I cannot find any business use case where we would want an ordered_locations strategy

guewen replied:
Put in the first bin in the row in front of you everytime (put in the bin closest to you). As you have constraints, it will take the next available in the row.

jbaudoux replied:
I just cannot figure out who would like to work like that ;-) Not an issue of course. Just a side note.

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+        if category and not product:
+            product = quant.product_id
+        for rule in putaway_rules:
+            dest_location = rule._get_destination_location()
```
Either dest_location is a bin (fixed location strategy), or is an area (chaotic strategy).
In case of area, the storage type must be defined on the area location.

So we can check here "dest_location" is allowed for the storage type. Otherwise, jump to next rule.

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+            storage_locations = dest_location.get_storage_locations(
+                product=product
+            )
+            allowed_location = storage_locations.select_first_allowed_location(
```
No need to check locations one by one. "Area" contains the allowed storage type.

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+        )
+        if category and not product:
+            product = quant.product_id
+        for rule in putaway_rules:
```
There is a design issue. Put-away rules are fixed location rules, and storage strategy is dynamic/chaotic location rule. They must work independently.
When you set a chaotic storage strategy on an area, it must be applied even if you do not have a put-away rule.
You must first apply the put-away rules to find the dest location (standard odoo), then you must apply the storage strategy.

guewen replied:
> it must be applied even if you do not have a put-away rule

why?

jbaudoux replied:
Take an area filled in by internal replenishment (so no put-away rule from reception). You could still want to apply chaotic on the area.

guewen replied:
As a temporary work-around, this commit allows this use case to work: https://github.com/OCA/wms/pull/5/commits/ffc6291a8cc3b326727ed0a93c0cdc1c611f67e8

Later, we'll probably want to take the final location returned by `StockLocation.get_putaway_strategy` and apply the storage type strategy on it.

One notable difference though, with the current behavior, you can have 2 rules as:

* Stock → Shelving
* Stock → Overflow

If it doesn't find any possible bin in Shelving, it will try in Overflow.
If we do the change we discuss, no place in Shelving means no put-away possible. Which in maybe not an issue in practice, I don't know.

jbaudoux commented on stock_putaway_storage_type_strategy/models/stock_location.py:
```diff
@@ -0,0 +1,166 @@
+
+    def _get_ordered_children_locations(self):
+        return self.search(
+            [('id', 'child_of', self.ids), ('id', '!=', self.id)]
```
We need to specify the wanted depth. If you have a storage strategy on an area having sub-areas and then bins, then I see 2 solutions:
* you only want to select bins. That's most likely where you will put the goods
* you only want to select a direct child (a sub-area). Then possibly you also have a storage strategy on the sub-area that selects a bin. That offers more flexibility but probably to specific
I would go for solution 1.

guewen replied:
In 2 you are speaking about the "recursive" strategies that we decided not to support with @jgrandguillaume , we only want bins.

jbaudoux reviewed (changes requested):
Additionally, there is an issue when moving complete pack. As put-away is computed per quant, potentially you can end-up with quants sent to different locations, especially in case of chaotic storage. I can imagine a single product pack with multiple lots.

Commit:
Add storage type on quant package

We need to be able to set the storage type manually, for instance
when the quantity is not equal to the product packaging (open pallet).

* On stock.quant.package, we have a new field product_packaging_id
  instead of using packaging_id, because the latter is for delivery
  packaging
* When the product_packaging_id is set on quant package, it
  automatically sets the storage type

Commit:
Relax the domain on put-away rule destination

Currently, we may need to have a source going to the same destination.
Example: WER → WER

It's a bit weird but the reason is this: if we want to apply a chaotic
strategy when we move to WER, as the strategy is applied on a put-away
rule's destination, we need such put-away rule.

Later, we may want to decorrelate the standard put-away rules and the
storage type strategy: apply the put-away rule first to the destination,
then apply the storage type strategy of this location.

grindtildeath commented:
Replaced by #8
